src/main/scala/org/ensime/indexer/FileWatchers.scala

```diff
@@ -57,12 +53,61 @@ class ClassfileWatcher(
   fm.setRecursive(true)
   fm.start()
 
-  fm.addFile(config.root)
+  // WORKAROUND https://issues.apache.org/jira/browse/VFS-536
+  // We don't have a dedicated test for this because it is an upstream bug
+  private val workaround = new DefaultFileMonitor(new FileListener {
+    private def targets =
+      config.compileClasspath.filter(_.isDirectory).map(vfile).map(_.getName)
+    def watched(event: FileChangeEvent) = {
+      val dir = event.getFile
+      val name = dir.getName
+      targets.exists(name.isAncestor(_))
+    }
+    def fileChanged(event: FileChangeEvent): Unit =
+      if (watched(event)) reset()
+    def fileCreated(event: FileChangeEvent): Unit =
+      if (watched(event)) reset()
+    def fileDeleted(event: FileChangeEvent): Unit = {}
+  })
+
+  workaround.setRecursive(false)
+  workaround.start()
+
+  private def ancestors(f: File): List[File] = {
+    val parent = f.getParentFile()
+    if (parent == null) Nil
+    else parent :: ancestors(parent)
+  }
+
+  // If directories are recreated, triggering the VFS-536 bug, we end up
+  // calling reset() a lot of times. We should probably debounce it, but
+  // it seems to happen so quickly that no damage is done.
+  private def reset(): Unit = {
+    log.info("Setting up new file watchers")
+    // must remove then add to avoid leaks
+
+    for {
+      d <- config.targetClasspath
+    } {
+      fm.removeFile(d)
+      fm.addFile(d)
+
+      // for-comp inside for-comp madness, w00t!
```

Comment:
nope, you can't chain that sort of thing (it would have been nice, because it would work for the whole block)